Can't build from source with crystal installed from snap

I followed [Install from sources](https://crystal-lang.org/install/from_sources/) with [Crystal installed from snap](https://crystal-lang.org/install/on_ubuntu/#snapcraft).

```bash
$ make
Using /usr/bin/llvm-config-15 [version= 15.0.7]
g++ -c  -o src/llvm/ext/llvm_ext.o src/llvm/ext/llvm_ext.cc -I/usr/lib/llvm-15/include -std=c++14   -fno-exceptions -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS
CRYSTAL_CONFIG_BUILD_COMMIT="df960cf76" CRYSTAL_CONFIG_PATH='$ORIGIN/../share/crystal/src' SOURCE_DATE_EPOCH="1687369226" CC="cc -fuse-ld=lld" CRYSTAL_CONFIG_LIBRARY_PATH='$ORIGIN/../lib/crystal' ./bin/crystal build -D strict_multi_assign -D preview_overload_order -Dwithout_interpreter  -o .build/crystal src/compiler/crystal.cr -D without_openssl -D without_zlib -D use_pcre2
error: unknown command "build", see 'snap help'.
make: *** [Makefile:214: .build/crystal] Error 64
```

The issue is that the `realpath` command for the following line

```sh
PARENT_CRYSTAL_COMMAND=$(realpath "$(command -v "$PARENT_CRYSTAL" 2> /dev/null)" 2> /dev/null)
```

resolves to `/usr/bin/snap`.

The following rough patch for the binstub resolves the issue:

```patch
diff --git a/bin/crystal b/bin/crystal
index 8e42d4e7e..800c1277c 100755
--- a/bin/crystal
+++ b/bin/crystal
@@ -154,6 +154,9 @@ if [ -z "${PARENT_CRYSTAL##*/*}" -a "$(realpath "$PARENT_CRYSTAL")" = "$SCRIPT_P
 fi
 
 PARENT_CRYSTAL_COMMAND=$(realpath "$(command -v "$PARENT_CRYSTAL" 2> /dev/null)" 2> /dev/null)
+if [ $(basename "$PARENT_CRYSTAL_COMMAND") = "snap" ]; then
+  PARENT_CRYSTAL_COMMAND="$(command -v "$PARENT_CRYSTAL" 2> /dev/null)"
+fi
 PARENT_CRYSTAL_EXISTS=$(test !$?)
 
 # check if the parent crystal command refers to this script

```

There's probably a more proper way to handle the snap case. If that's good enough I'm happy to submit a PR. Otherwise I'm open to suggestions on how to improve the patch.

--- a/bin/crystal
+++ b/bin/crystal
@@ -153,12 +153,8 @@ if [ -z "${PARENT_CRYSTAL##*/*}" -a "$(realpath "$PARENT_CRYSTAL")" = "$SCRIPT_P
   PARENT_CRYSTAL="crystal"
 fi
 
-PARENT_CRYSTAL_COMMAND=$(command -v "$PARENT_CRYSTAL" 2> /dev/null)
-PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARENT_CRYSTAL_COMMAND" 2> /dev/null)
-PARENT_CRYSTAL_EXISTS=$(test !$?)
-
 # check if the parent crystal command refers to this script
-if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
+if [ "$(realpath "$(command -v "$PARENT_CRYSTAL" 2> /dev/null)" 2> /dev/null)" = "$SCRIPT_PATH" ]; then
   # remove the path to this script from PATH
   NEW_PATH="$(remove_path_item "$(remove_path_item "$PATH" "$SCRIPT_ROOT")" "bin")"
   # if the PATH did not change it will lead to an infinite recursion => display error
@@ -166,18 +162,18 @@ if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
     __error_msg 'Could not remove the script bin/crystal from the PATH. Remove it by hand or set the CRYSTAL env variable'
     exit 1
   fi
-  PARENT_CRYSTAL_COMMAND=$(PATH=$NEW_PATH command -v "$PARENT_CRYSTAL" 2> /dev/null)
-  PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARENT_CRYSTAL_COMMAND" 2> /dev/null)
-  PARENT_CRYSTAL_EXISTS=$(test !$?)
-  if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
+  PARENT_CRYSTAL=$(PATH=$NEW_PATH command -v "$PARENT_CRYSTAL" 2> /dev/null)
+  if [ "$(realpath "$PARENT_CRYSTAL" 2> /dev/null)" = "$SCRIPT_PATH" ]; then
     __error_msg 'Could not remove the script bin/crystal from the PATH. Remove it by hand or set the CRYSTAL env variable'
     exit 1
   fi
 fi
 
+command -v "$PARENT_CRYSTAL" > /dev/null 2> /dev/null
+PARENT_CRYSTAL_EXISTS=$(test !$?)
 if ($PARENT_CRYSTAL_EXISTS); then
   if [ -z "$CRYSTAL_CONFIG_LIBRARY_PATH" ] || [ -z "$CRYSTAL_LIBRARY_PATH" ]; then
-    CRYSTAL_INSTALLED_LIBRARY_PATH="$($PARENT_CRYSTAL_COMMAND env CRYSTAL_LIBRARY_PATH 2> /dev/null || echo "")"
+    CRYSTAL_INSTALLED_LIBRARY_PATH="$($PARENT_CRYSTAL env CRYSTAL_LIBRARY_PATH 2> /dev/null || echo "")"
     export CRYSTAL_LIBRARY_PATH=${CRYSTAL_LIBRARY_PATH:-$CRYSTAL_INSTALLED_LIBRARY_PATH}
     export CRYSTAL_CONFIG_LIBRARY_PATH=${CRYSTAL_CONFIG_LIBRARY_PATH:-$CRYSTAL_INSTALLED_LIBRARY_PATH}
   fi
@@ -195,5 +191,5 @@ elif !($PARENT_CRYSTAL_EXISTS); then
   __error_msg 'You need to have a crystal executable in your path! or set CRYSTAL env variable'
   exit 1
 else
-  exec "$PARENT_CRYSTAL_COMMAND" "$@"
+  exec "$PARENT_CRYSTAL" "$@"
 fi

--- a/bin/crystal
+++ b/bin/crystal
@@ -153,11 +153,12 @@ if [ -z "${PARENT_CRYSTAL##*/*}" -a "$(realpath "$PARENT_CRYSTAL")" = "$SCRIPT_P
   PARENT_CRYSTAL="crystal"
 fi
 
-PARENT_CRYSTAL_COMMAND=$(realpath "$(command -v "$PARENT_CRYSTAL" 2> /dev/null)" 2> /dev/null)
+PARENT_CRYSTAL_COMMAND=$(command -v "$PARENT_CRYSTAL" 2> /dev/null)
+PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARNET_CRYSTAL_COMMAND" 2> /dev/null)
 PARENT_CRYSTAL_EXISTS=$(test !$?)
 
 # check if the parent crystal command refers to this script
-if [ "$PARENT_CRYSTAL_COMMAND" = "$SCRIPT_PATH" ]; then
+if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
   # remove the path to this script from PATH
   NEW_PATH="$(remove_path_item "$(remove_path_item "$PATH" "$SCRIPT_ROOT")" "bin")"
   # if the PATH did not change it will lead to an infinite recursion => display error

--- a/bin/crystal
+++ b/bin/crystal
@@ -166,9 +166,10 @@ if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
     __error_msg 'Could not remove the script bin/crystal from the PATH. Remove it by hand or set the CRYSTAL env variable'
     exit 1
   fi
-  PARENT_CRYSTAL_COMMAND=$(realpath "$(PATH=$NEW_PATH command -v "$PARENT_CRYSTAL" 2> /dev/null)" 2> /dev/null)
+  PARENT_CRYSTAL_COMMAND=$(PATH=$NEW_PATH command -v "$PARENT_CRYSTAL" 2> /dev/null)
+  PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARENT_CRYSTAL_COMMAND" 2> /dev/null)
   PARENT_CRYSTAL_EXISTS=$(test !$?)
-  if [ "$PARENT_CRYSTAL_COMMAND" = "$SCRIPT_PATH" ]; then
+  if [ "$PARENT_CRYSTAL_COMMAND_REALPATH" = "$SCRIPT_PATH" ]; then
     __error_msg 'Could not remove the script bin/crystal from the PATH. Remove it by hand or set the CRYSTAL env variable'
     exit 1
   fi

--- a/bin/crystal
+++ b/bin/crystal
@@ -154,7 +154,7 @@ if [ -z "${PARENT_CRYSTAL##*/*}" -a "$(realpath "$PARENT_CRYSTAL")" = "$SCRIPT_P
 fi
 
 PARENT_CRYSTAL_COMMAND=$(command -v "$PARENT_CRYSTAL" 2> /dev/null)
-PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARNET_CRYSTAL_COMMAND" 2> /dev/null)
+PARENT_CRYSTAL_COMMAND_REALPATH=$(realpath "$PARENT_CRYSTAL_COMMAND" 2> /dev/null)
 PARENT_CRYSTAL_EXISTS=$(test !$?)
 
 # check if the parent crystal command refers to this script

